### ISSUE TITLE
Problem with decoding Datamatrix barcode

### ISSUE THREAD

**Issue by OL2015** [zxing-cpp/zxing-cpp#669]

**Comment by axxel** [comment]
The problem is the upper right corner of the symbol. There is a local vertical stretching or deformation. The current approach of placing a regular sampling grid over the symbol fails in that case. You might want to have look at #252.

**Comment by OL2015** [comment]
Thank you for the answer. I made stretching corrections based on timing marks, and as a result, the image was successfully decoded. 

May I suggest adding axis stretching correction to the ZXing library? Vertical stretching can be caused by barcode printers, and this is not just a single occurrence.

Anycase, I'll use image stretching if ZXing returns matrix position, but zero-lencht text.

**Comment by axxel** [comment]
Sorry for the delayed response. I somehow missed your last comment. Can you explain to me what you mean by

> stretching corrections based on timing marks

and

> adding axis stretching correction to the ZXing library

**Comment by OL2015** [comment]
I just discovered that my response sent via email did not appear in the thread, so I will repeat it here.
The attached image gives an illustration. 
 
![unnamed](https://github.com/zxing-cpp/zxing-cpp/assets/13064569/9c9ee5b5-5dbc-4083-a092-2450d0bb6461)

The DataMatrix fragment on the left exhibits a distortion along the Y-axis, which is particularly noticeable in the first three rows. This type of distortion can be corrected as the number of timings can be calculated and the dimensions of the image are known. Due to the irregular nature of the distortion, I perform the correction on a row-by-row basis.

The likely cause of the Y-axis distortion is an eccentricity in the barcode printer or some other mechanical issue. While I'm not certain how common this problem is, if it proves to be a frequent issue, incorporating the correction directly into the library could be a viable solution.

**Comment by axxel** [comment]
Thanks for following up on this but...

> the number of timings can be calculated

again, I don't know what you mean by that statement. Can you point me to some source code with your changes?

**Comment by OL2015** [comment]
Yes, I can send the code snippets to you. Could you please provide your email address where I should send them?

**Comment by axxel** [comment]
zxingcpp at gmail.com

**Comment by OL2015** [comment]
I sent the code to you mail. Please find it.

**Comment by Jakub444** [comment]
Hello, I have the same problem with this photo.  Another program reads it without a problem.  Is there such a solution?
![20240412_114930](https://github.com/zxing-cpp/zxing-cpp/assets/52671558/7566ac57-e5e6-49d0-bdd8-1ea7ffea9f59)

**Comment by Sec-ant** [comment]
I think this is also fixed by https://github.com/zxing-cpp/zxing-cpp/commit/b304f6656b865be11b00741d430b7d0027bc0ab4. I tested the image posted by OP using the master branch code, it detects well. cc @axxel

**Comment by axxel** [comment]
@Sec-ant you are right, the first one from the OP works now but the most recent sample from @Jakub444 still needs work to be done. So I let this stay open for now.

**Comment by axxel** [comment]
My recent work considerably improved the situation. The second image can actually decode now if the still experimental `tryDenoise(true)` flag is set. The alignment pattern evaluation improvements I implemented help only if the finder-pattern is fixed (small discontinuation in lower leg of the L-shape), see #234. So technically, the current HEAD still can't decode this image as is.
